SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverRPC.cpp

```diff
@@ -0,0 +1,471 @@
+{
+	const USpatialGDKSettings* Settings = GetDefault<USpatialGDKSettings>();
+
+	for (auto& Update : Updates)
+	{
+		FSpatialGDKSpanId SpanId;
+		if (EventTracer != nullptr)
+		{
+			SpanId = EventTracer->TraceEvent(FSpatialTraceEventBuilder::CreateMergeSendRPCs(Update.EntityId, Update.Update.component_id),
+											 /* Causes */ Update.Spans.GetData()->GetConstId(), /* NumCauses */ Update.Spans.Num());
+		}
+		NetDriver->Connection->SendComponentUpdate(Update.EntityId, &Update.Update, SpanId);
+	}
+
+	if (Updates.Num() > 0 && Settings->bWorkerFlushAfterOutgoingNetworkOp)
+	{
+		NetDriver->Connection->Flush();
+	}
+}
+
+bool USpatialNetDriverRPC::CanExtractRPC(Worker_EntityId EntityId)
+{
+	const TWeakObjectPtr<UObject> ActorReceivingRPC = NetDriver->PackageMap->GetObjectFromEntityId(EntityId);
+	if (!ActorReceivingRPC.IsValid())
+	{
+		UE_LOG(LogSpatialRPCService, Log,
+			   TEXT("Entity receiving ring buffer RPC does not exist in PackageMap, possibly due to corresponding actor getting "
+					"destroyed. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+	return true;
+}
+
+bool USpatialNetDriverRPC::CanExtractRPCOnServer(Worker_EntityId EntityId)
+{
+	const TWeakObjectPtr<UObject> ActorReceivingRPC = NetDriver->PackageMap->GetObjectFromEntityId(EntityId);
+	if (!ActorReceivingRPC.IsValid())
+	{
+		UE_LOG(LogSpatialRPCService, Log,
+			   TEXT("Entity receiving ring buffer RPC does not exist in PackageMap, possibly due to corresponding actor getting "
+					"destroyed. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+
+	const bool bActorRoleIsSimulatedProxy = Cast<AActor>(ActorReceivingRPC.Get())->Role == ROLE_SimulatedProxy;
+	if (bActorRoleIsSimulatedProxy)
+	{
+		UE_LOG(LogSpatialRPCService, Verbose,
+			   TEXT("Will not process server RPC, Actor role changed to SimulatedProxy. This happens on migration. Entity: %lld"),
+			   EntityId);
+		return false;
+	}
+	return true;
+}
+
+struct RAIIParamsHolder
+{
+	RAIIParamsHolder(UFunction* InFunction, uint8* Memory)
+		: Function(InFunction)
+	{
+		Parms = Memory;
+		FMemory::Memzero(Parms, Function->ParmsSize);
+	}
+
+	~RAIIParamsHolder()
+	{
+		// Destroy the parameters.
+		// warning: highly dependent on UObject::ProcessEvent freeing of parms!
+		for (TFieldIterator<GDK_PROPERTY(Property)> It(Function); It && It->HasAnyPropertyFlags(CPF_Parm); ++It)
+		{
+			It->DestroyValue_InContainer(Parms);
+		}
+	}
+
+	UFunction* Function;
+	uint8* Parms;
+};
+
+bool USpatialNetDriverRPC::ApplyRPC(Worker_EntityId EntityId, SpatialGDK::ReceivedRPC RPCData, const FRPCMetaData& MetaData)
+{
+	constexpr bool RPCConsumed = true;
+
+	FUnrealObjectRef ObjectRef(EntityId, RPCData.Offset);
+	TWeakObjectPtr<UObject> TargetObjectWeakPtr = NetDriver->PackageMap->GetObjectFromUnrealObjectRef(ObjectRef);
+	UObject* TargetObject = TargetObjectWeakPtr.Get();
+
+	const FClassInfo& ClassInfo = NetDriver->ClassInfoManager->GetOrCreateClassInfoByObject(TargetObjectWeakPtr.Get());
```

Comment:
Already done the Get() above?

Reply:
I assume NetDriver and/or ClassInfoManager cannot be null?

Reply:
Note - GetOrCreateClassInfoByObject will access a nullptr if provided one (it doesn't verify that its invalid), and neither does this function?